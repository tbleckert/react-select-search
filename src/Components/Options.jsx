import {
    memo,
    useCallback,
    useEffect,
    useRef,
} from 'react';
import PropTypes from 'prop-types';
import { optionType, valueType } from '../types';
import OptionsList from './OptionsList';

const Options = ({
    options,
    optionProps,
    snapshot,
    cls,
    renderGroupHeader,
    renderOption,
    emptyMessage,
}) => {
    const selectRef = useRef(null);
    const { value, highlighted } = snapshot;
    const renderEmptyMessage = useCallback(() => {
        if (emptyMessage === null) {
            return null;
        }

        return (
            <li className={cls('not-found')}>
                {(typeof emptyMessage === 'function') ? emptyMessage() : emptyMessage}
            </li>
        );
    }, [emptyMessage, cls]);

    useEffect(() => {
        const { current } = selectRef;

        if (!current || (highlighted < 0 && value === null)) {
            return;
        }

        const val = (Array.isArray(value)) ? value[0] : value;
        const query = (highlighted > -1) ? `[data-index="${highlighted}"]` : `[data-value="${escape(val)}"]`;
        const selected = current.querySelector(query);

        if (selected) {
            const rect = current.getBoundingClientRect();
            const selectedRect = selected.getBoundingClientRect();

            current.scrollTop = selected.offsetTop - (rect.height / 2) + (selectedRect.height / 2);
        }
    }, [value, highlighted, selectRef]);

    return (
        // eslint-disable-next-line jsx-a11y/no-static-element-interactions
        <div className={cls('select')} ref={selectRef} onMouseDown={(e) => e.preventDefault()}>
            {options.length ? (
                <OptionsList
                    optionProps={optionProps}
                    snapshot={snapshot}
                    options={options}
                    renderOption={renderOption}
                    renderGroupHeader={renderGroupHeader}
                    cls={cls}
                />
            ) : (
                <ul className={cls('options')}>
                    {renderEmptyMessage()}
                </ul>
            )}
        </div>
    );
};

Options.defaultProps = {
    renderOption: null,
    renderGroupHeader: null,
    emptyMessage: null,
};

Options.propTypes = {
    options: PropTypes.arrayOf(optionType).isRequired,
    optionProps: PropTypes.shape({
        tabIndex: PropTypes.string.isRequired,
        onMouseDown: PropTypes.func.isRequired,
    }).isRequired,
    snapshot: PropTypes.shape({
        highlighted: PropTypes.number.isRequired,
        value: valueType,
    }).isRequired,
    cls: PropTypes.func.isRequired,
    renderGroupHeader: PropTypes.func,
    renderOption: PropTypes.func,
    emptyMessage: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.func,
    ]),
};

export default memo(Options);
