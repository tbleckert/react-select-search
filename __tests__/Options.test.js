import renderer from 'react-test-renderer';
import Options from '../src/Components/Options';
import { countries } from './data';

describe('Test Options component', () => {
    test('Not found', () => {
        const component = renderer.create(
            <Options
                options={[]}
                cls={(cls) => cls}
                emptyMessage="Not found"
                optionProps={{
                    tabIndex: '0',
                    onMouseDown: () => {},
                }}
                snapshot={{
                    highlighted: -1,
                    option: null,
                    focus: false,
                    value: null,
                }}
            />
        );

        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();

        expect(component.root.findAllByType('li').length).toBe(1);
        expect(component.root.findByType('li').props.className).toBe('not-found');
        expect(component.root.findByType('li').props.children).toBe('Not found');
    });
});
