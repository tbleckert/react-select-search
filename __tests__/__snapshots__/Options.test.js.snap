// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Options component Not found 1`] = `
<div
  className="select"
  onMouseDown={[Function]}
>
  <ul
    className="options"
  >
    <li
      className="not-found"
    >
      Not found
    </li>
  </ul>
</div>
`;
